pkg/objstore/cos/cos.go

```diff
@@ -0,0 +1,416 @@
+package cos
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"math/rand"
+	"net/http"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/improbable-eng/thanos/pkg/objstore"
+	"github.com/improbable-eng/thanos/pkg/runutil"
+	"github.com/mozillazg/go-cos"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"gopkg.in/alecthomas/kingpin.v2"
+)
+
+const (
+	opObjectsList  = "ListBucket"
+	opObjectInsert = "PutObject"
+	opObjectGet    = "GetObject"
+	opObjectHead   = "HEADObject"
+	opObjectDelete = "DeleteObject"
+)
+
+// DirDelim is the delimiter used to model a directory structure in an object store bucket.
+const DirDelim = "/"
+
+// Bucket implements the store.Bucket interface against cos-compatible(Tencent Object Storage) APIs.
+type Bucket struct {
+	logger   log.Logger
+	client   *cos.Client
+	opsTotal *prometheus.CounterVec
+}
+
+// Config encapsulates the necessary config values to instantiate an cos client.
+type Config struct {
+	Bucket    string
+	AppId     string
+	Region    string
+	SecretId  string
+	SecretKey string
+}
+
+// RegisterCosParams registers the cos flags and returns an initialized Config struct.
+func RegisterCosParams(cmd *kingpin.CmdClause) *Config {
+	var cosConfig Config
+
+	cmd.Flag("cos.bucket", "Cos-Compatible API bucket name for stored blocks.").
+		PlaceHolder("<bucket>").Envar("COS_BUCKET").StringVar(&cosConfig.Bucket)
+
+	cmd.Flag("cos.appid", "Cos-Compatible API endpoint for stored blocks.").
+		PlaceHolder("<appid>").Envar("COS_APPID").StringVar(&cosConfig.AppId)
+
+	cmd.Flag("cos.region", "Cos-Compatible API region for stored blocks.").
+		PlaceHolder("<region>").Envar("COS_REGION").StringVar(&cosConfig.Region)
+
+	cosConfig.SecretKey = os.Getenv("COS_SECRET_KEY")
```

Comment:
Do you think we can actually reuse existing flags from S3. Somehow extract those into package that will be used for all providers. Seems like we could possibly find some set of flags that will be compatible for all providers. Maybe not in this PR, but it starts to worry (: